core/domain/exp_services.py

```diff
@@ -455,78 +455,123 @@ def apply_change_list(
     """
     exploration = exp_fetchers.get_exploration_by_id(exploration_id)
     try:
         to_param_domain = param_domain.ParamChange.from_dict
         for change in change_list:
             if change.cmd == exp_domain.CMD_ADD_STATE:
-                exploration.add_states([change.state_name])
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                add_state_cmd = cast(
+                    exp_domain.AddExplorationStateCmd,
+                    change
+                )
+                exploration.add_states([add_state_cmd.state_name])
             elif change.cmd == exp_domain.CMD_RENAME_STATE:
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                rename_state_cmd = cast(
+                    exp_domain.RenameExplorationStateCmd,
+                    change
+                )
                 exploration.rename_state(
-                    change.old_state_name, change.new_state_name)
+                    rename_state_cmd.old_state_name,
+                    rename_state_cmd.new_state_name
+                )
             elif change.cmd == exp_domain.CMD_DELETE_STATE:
-                exploration.delete_state(change.state_name)
+                # Here we use cast because we are narrowing down the type from
+                # ExplorationChange to a specific change command.
+                delete_state_cmd = cast(
+                    exp_domain.DeleteExplorationStateCmd,
+                    change
+                )
+                exploration.delete_state(delete_state_cmd.state_name)
             elif change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY:
                 state: state_domain.State = exploration.states[
                     change.state_name]
                 if (change.property_name ==
                         exp_domain.STATE_PROPERTY_PARAM_CHANGES):
-                    # Here change is an instance of ExplorationChange and every
-                    # attribute on ExplorationChange is defined dynamically, and
-                    # every dynamically defined attribute have str type. Thus to
-                    # convert the type to list, we have used assert here.
-                    assert isinstance(change.new_value, list)
+                    # Here we use cast because this 'if' condition forces
+                    # change to have type EditExpStatePropertyParamChangesCmd.
+                    edit_param_changes_cmd = cast(
+                        exp_domain.EditExpStatePropertyParamChangesCmd,
+                        change
+                    )
                     state.update_param_changes(list(map(
-                            to_param_domain, change.new_value)))
+                            to_param_domain, edit_param_changes_cmd.new_value)
```

Comment:
Indent by 4 only, not 8. Then you can do:

```
    state.update_param_changes(list(map(
        to_param_domain, edit_param_changes_cmd.new_value
    )))
```

Reply:
Done